lib/improver/wind_downscaling.py

```diff
@@ -0,0 +1,967 @@
+        if hasattr(modoro_cube, "operation"):
+            modoro_cube = modoro_cube.operation()
+        if hasattr(z0_cube, "operation"):
+            z0_cube = z0_cube.operation()
+        if hasattr(height_levels_cube, "operation"):
+            height_levels_cube = height_levels_cube.operation()
+
+        x_name, y_name, _, _ = self.find_coord_names(pporo_cube)
+        # Some checking that all the grids match
+        if not (self.check_ancils(a_over_s_cube, sigma_cube, z0_cube,
+                                  pporo_cube, modoro_cube)):
+            raise ValueError("ancillary grids are not consistent")
+        # I want this ordering. Careful: iris.cube.Cube.slices is unreliable.
+        self.a_over_s = next(a_over_s_cube.slices([y_name, x_name]))
+        self.sigma = next(sigma_cube.slices([y_name, x_name]))
+        try:
+            self.z_0 = next(z0_cube.slices([y_name, x_name]))
+        except AttributeError:
+            self.z_0 = z0_cube
+        except Exception as exc:
+            emsg = "'{0}' while z0 setting. Arguments '{1}'."
+            raise ValueError(emsg.format(exc.message, exc.args))
+        self.pp_oro = next(pporo_cube.slices([y_name, x_name]))
+        self.model_oro = next(modoro_cube.slices([y_name, x_name]))
+        self.ppres = self.calc_av_ppgrid_res(pporo_cube)
+        self.modres = modres
+        self.height_levels = height_levels_cube
+        if z0_cube is None:
+            self.l_no_winddownscale = True
+        else:
+            self.l_no_winddownscale = False
+        self.x_name = None
+        self.y_name = None
+        self.z_name = None
+        self.t_name = None
+
+    def find_coord_names(self, cube):
+        """Extract x, y, z, and time coordinate names.
+
+        Parameters:
+        ----------
+        cube: cube
+            some iris cube to find coordinate names from
+
+        Returns:
+        -------
+        xname: str
+            name of the axis name in x-direction
+        yname: str
+            name of the axis name in y-direction
+        zname: str
+            name of the axis name in z-direction
+        tname: str
+            name of the axis name in t-direction
+
+        """
+        clist = set([cube.coords()[iii].name() for iii in
+                     range(len(cube.coords()))])
+        try:
+            xname = list(clist.intersection(self.xcoordnames))[0]
+        except IndexError:
+            xname = None
+        except Exception as exc:
+            print("'{0}' while xname setting. Args: {1}.".format(exc.message,
+                                                                 exc.args))
+        try:
+            yname = list(clist.intersection(self.ycoordnames))[0]
+        except IndexError:
+            yname = None
+        except Exception as exc:
+            print("'{0}' while yname setting. Args: {1}.".format(exc.message,
+                                                                 exc.args))
+        try:
```

Comment:
Not currently replaced, as there could be alternative valid names for the z-coordinate.